libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -733,48 +733,53 @@ void GCS_MAVLINK::send_message(enum ap_message id)
         save_signing_timestamp(false);
     }
     
-    // see if we can send the deferred messages, if any
+    // see if we can send the deferred messages, if any:
     while (num_deferred_messages != 0) {
         if (!try_send_message(deferred_messages[next_deferred_message])) {
             break;
         }
         next_deferred_message++;
-        if (next_deferred_message == MSG_RETRY_DEFERRED) {
+        if (next_deferred_message == ARRAY_SIZE(deferred_messages)) {
             next_deferred_message = 0;
         }
         num_deferred_messages--;
     }
 
+    // MSG_RETRY_DEFERRED is a "marker" message only used to push the queue
     if (id == MSG_RETRY_DEFERRED) {
         return;
     }
 
-    // this message id might already be deferred
+    // if there are no deferred messages, attempt to send straight away:
+    if (num_deferred_messages == 0) {
+        if (try_send_message(id)) {
+            // yay, we sent it!
+            return;
+        }
+    }
+
+    // we failed to send the message this time around, so try to defer:
+    if (num_deferred_messages == ARRAY_SIZE(deferred_messages)) {
+        // the defer buffer is full, discard this attempt to send.
+        // Note that the message *may* already be in the defer buffer
```

Comment:
Again, we currently size it so you can defer one of each message (and at 4bytes*~20 elements, *shrug*).  But it doesn't need to be that way.